core/object/object.h

```diff
@@ -47,103 +47,471 @@ template <typename T>
+	// The hint string is the placeholder text to use.
+	PROPERTY_HINT_PLACEHOLDER_TEXT,
+
+	// Hints that a `Color` property should be edited without affecting its transparency
+	// (`Color.a` is not editable).
+	PROPERTY_HINT_COLOR_NO_ALPHA,
+
+	// Hints that the property's value is an object encoded as object ID,
+	// with its type specified in the hint string. Used by the debugger.
 	PROPERTY_HINT_OBJECT_ID,
-	PROPERTY_HINT_TYPE_STRING, ///< a type string, the hint is the base type to choose
-	PROPERTY_HINT_NODE_PATH_TO_EDITED_NODE, // Deprecated.
-	PROPERTY_HINT_OBJECT_TOO_BIG, ///< object is too big to send
+
+	// If a property is `String`, hints that the property represents a particular type (class).
+	// This allows to select a type from the create dialog. The property will store the selected type as a string.
+	// If a property is `Array`, hints the editor how to show elements.
+	// The `hint_string` must encode nested types using `":"` and `"/"`.
+	PROPERTY_HINT_TYPE_STRING,
+
+	// Deprecated. This hint is not used by the engine.
+	PROPERTY_HINT_NODE_PATH_TO_EDITED_NODE,
+
+	// Hints that an object is too big to be sent via the debugger.
+	PROPERTY_HINT_OBJECT_TOO_BIG,
+
+	// Hints that the hint string specifies valid node types for property of type `NodePath`.
+	// The hint string can specify multiple classes separated by commas (e.g. `"Node2D,Node3D"`);
+	// all subclasses are included automatically.
+	// Editing the property will show a dialog for picking a node from the scene.
+	// This is similar to `PROPERTY_HINT_RESOURCE_TYPE`, but for `NodePath` properties.
+	// Use `PROPERTY_HINT_NODE_TYPE` for `Node` properties instead.
 	PROPERTY_HINT_NODE_PATH_VALID_TYPES,
-	PROPERTY_HINT_SAVE_FILE, ///< a file path must be passed, hint_text (optionally) is a filter "*.png,*.wav,*.doc,". This opens a save dialog
-	PROPERTY_HINT_GLOBAL_SAVE_FILE, ///< a file path must be passed, hint_text (optionally) is a filter "*.png,*.wav,*.doc,". This opens a save dialog
-	PROPERTY_HINT_INT_IS_OBJECTID, // Deprecated.
+
+	// Hints that a `String` property is a path to a file.
+	// Editing it will show a file dialog for picking the path for the file to be saved at.
+	// The dialog has access to the project's directory.
+	// The hint string can be a set of filters with wildcards like `"*.png,*.jpg"`. See also `PROPERTY_HINT_FILE`.
+	PROPERTY_HINT_SAVE_FILE,
+
+	// Hints that a `String` property is a path to a file.
+	// Editing it will show a file dialog for picking the path for the file to be saved at.
+	// The dialog has access to the entire filesystem.
+	// The hint string can be a set of filters with wildcards like `"*.png,*.jpg"`. See also `PROPERTY_HINT_GLOBAL_FILE`.
+	PROPERTY_HINT_GLOBAL_SAVE_FILE,
+
+	// Deprecated.
+	PROPERTY_HINT_INT_IS_OBJECTID,
+
+	// Hints that an `int` property is a pointer. Used by GDExtension.
 	PROPERTY_HINT_INT_IS_POINTER,
+
+	// Hints that a property is an [Array] with the stored type specified in the hint string.
```

Comment:
This should mention `MAKE_RESOURCE_TYPE_HINT`, which is used to generate hint strings in most cases.

As well as full hit string format:
`vformat("%s/%s:%s", Variant::TYPE, PROPERTY_HINT_XXX, "hint_string")`.